### DISCUSSION TITLE
NextJS - syncing an anonymously signed in user to an existing user upon signing in

### DISCUSSION THREAD

**Discussion by sameerxanand** [supabase/supabase#22377] [Questions]
I'm creating an application that allows non-logged in users to read articles/content (up to 3) before being gated. Once gated, they can login using their email/OTP to resume browsing content. To implement this, I am thinking about using the new anonymous sign-in feature.

I'm confused about how to convert an anonymously signed in user/session to a verified one upon successfully authenticating (in my case, using email/OTP).

In my Next.js application, I have this `middleware.ts`:

```typescript
import { createMiddlewareClient } from '@supabase/auth-helpers-nextjs'
import { NextResponse } from 'next/server'

import type { NextRequest } from 'next/server'
import type { Database } from '@/util/types'

export async function middleware(req: NextRequest) {
  const res = NextResponse.next()

  // Create a Supabase client configured to use cookies
  const supabase = createMiddlewareClient<Database>({ req, res })

  // Refresh session if expired - required for Server Components
  const session = await supabase.auth.getSession()

  return res
}
```

Which I'll change to:

```typescript
import { createMiddlewareClient } from '@supabase/auth-helpers-nextjs'
import { NextResponse } from 'next/server'

import type { NextRequest } from 'next/server'
import type { Database } from '@/util/types'

export async function middleware(req: NextRequest) {
  const res = NextResponse.next()

  // Create a Supabase client configured to use cookies
  const supabase = createMiddlewareClient<Database>({ req, res })

  // Refresh session if expired - required for Server Components
  const session = await supabase.auth.getSession()

  if (!session.data.session) await supabase.auth.signInAnonymously()

  return res
}
```

I added the line:

```ts
if (!session.data.session) await supabase.auth.signInAnonymously()
```

To ensure all visitors on the site have a session, and the app can keep track of how many pages/pieces of content that session has accessed.

---

On my frontend, I have a signup server action connected with the login modal:

```ts
'use server'

import { z } from 'zod'
import { cookies } from 'next/headers'
import { createServerActionClient } from '@supabase/auth-helpers-nextjs'

const schema = z.object({
  email: z.string().email({ message: 'Please enter a valid email address.' }),
})

export async function signup(prevState: any, formData: FormData) {
  const validatedFields = schema.safeParse({
    email: formData.get('email'),
  })

  // Return early if the form data is invalid
  if (!validatedFields.success) {
    return {
      errors: validatedFields.error.flatten().fieldErrors,
    }
  }

  const supabase = createServerActionClient({ cookies })

  await supabase.auth.signInWithOtp({
    email: validatedFields.data.email,
    options: {
      shouldCreateUser: true,
    },
  })

  return { success: true, email: validatedFields.data.email }
}
```

Once this form is submitted, the UI changes to accept the 6-digit one-time code, and is verified using this server action:

```ts
'use server'

import { z } from 'zod'
import { cookies } from 'next/headers'
import { createServerActionClient } from '@supabase/auth-helpers-nextjs'
import { revalidatePath } from 'next/cache'

const schema = z.object({
  email: z.string().email({ message: 'Please enter a valid email address.' }),
  code: z.string().length(6),
})

export async function login(prevState: any, formData: FormData) {
  const validatedFields = schema.safeParse({
    email: formData.get('email'),
    code: formData.get('code'),
  })

  // Return early if the form data is invalid
  if (!validatedFields.success) {
    return {
      errors: validatedFields.error.flatten().fieldErrors,
    }
  }

  const supabase = createServerActionClient({ cookies })

  const response = await supabase.auth.verifyOtp({
    type: 'email',
    email: validatedFields.data.email,
    token: validatedFields.data.code,
  })

  if (!response.data.session || !response.data.user) {
    return {
      error: 'Invalid or expired login code.',
    }
  }

  revalidatePath('/')

  return { success: true }
}
```

However, signing-in with these server actions do not link that anonymous user with the now authenticated user.

The documentation says:

> https://supabase.com/docs/guides/auth/auth-anonymous
> You can use the [updateUser()](https://supabase.com/docs/reference/javascript/auth-updateuser) method to link an email or phone identity to the anonymous user.
> ```ts
> const { data, error } = await supabase.auth.updateUser({ email: 'example@email.com' })
> ```

But this isn't clear at what step I should do this. Calling `supabase.auth.signInWithOtp()` clears the anonymous session/user.

Is there a way to pass `supabase.auth.signInWithOtp()` and `supabase.auth.verifyOtp()` context of the anonymous user/session so that both get automatically linked upon signing in?

**Comment by truexor** [comment]
To link an anonymous user session with an authenticated user session in Supabase, you can follow these steps:

- sign in the user anonymously
- when the user chooses to log in with email/OTP, use `signInWithOtp()` to authenticate them
- once the user is authenticated successfully, use the `updateUser()` method to link the anonymous session with the authenticated session

```ts
// In your login server action
import { createServerActionClient } from '@supabase/auth-helpers-nextjs'
import { cookies } from 'next/headers'

export async function login(prevState: any, formData: FormData) {
  // Validate form data
  // ...

  const supabase = createServerActionClient({ cookies })

  // Sign in the user with OTP
  const response = await supabase.auth.signInWithOtp({
    email: validatedFields.data.email,
  })

  if (response.error) {
    return {
      error: response.error.message,
    }
  }

  // Update user to link anonymous session with authenticated session
  await supabase.auth.updateUser({ email: validatedFields.data.email })

  return { success: true }
}
```


By calling `updateUser()` after successfully authenticating the user with `signInWithOtp()`, you're linking the anonymous session with the authenticated session under the same user identity. This way, any data associated with the anonymous session will be retained and linked to the authenticated session.

**Reply by sameerxanand** [reply]
This doesn't work because calling `signInWithOtp()` immediately removes the existing anonymous session:

![CleanShot 2024-04-02 at 13 08 45@2x](https://github.com/supabase/supabase/assets/3168008/0264506c-d2b2-4d3d-ae34-4f1974002396)

Also, you wouldn't want to call `updateUser` until the verification step passes. `signInWithOtp()` just sends the user the code. You then have to call `verifyOtp()` to verify that the code the user provides matches what was sent to them in the email/text. Only then can you know that this user is really who they say they are.

**Reply by camwest** [reply]
From this guide: https://supabase.com/docs/guides/auth/auth-anonymous#link-an-email--phone-identity

I think you need to do:

```
const { data, error } = await supabase.auth.updateUser({ email: 'example@email.com' })

// verify the user's email by clicking on the email change link
// or entering the 6-digit OTP sent to the email address

// once the user has been verified, update the password
const { data, error } = await supabase.auth.updateUser({ password: 'password' })
```

**Reply by jkhaui** [reply]
So the anonymous sign-in feature essentially doesn't work at all then. Great

**Reply by silentworks** [reply]
@jkhaui I'm not sure how you read the solution provided by @camwest and still came to that conclusion. Have you tested what was stated? because it works for me.

**Reply by jkhaui** [reply]
> @jkhaui I'm not sure how you read the solution provided by @camwest and still came to that conclusion. Have you tested what was stated?

Yes that was the first thing I tried... it doesn't work for me. What happens is a new user is created but this isn't the same anonymous user that's carried over (the user id is different). I assume this is because - as @sameerxanand mentioned - `removeSession` is called immediately in `signInWithOtp`.

> because it works for me.

Great, are you able to please provide an example? I would appreciate it as I can't find any full-fledged demos of sign in with OTP + updating an anonymous user

**Reply by silentworks** [reply]
@jkhaui from your comment you aren't doing what @camwest suggested as you mentioned `signInWithOtp` rather than `updateUser`.

I've created an example repo here https://github.com/silentworks/supabase_anon_to_user and here is a video attached showing how this works with `updateUser` as @camwest described.

https://github.com/user-attachments/assets/3f2ff514-6eb9-4573-8508-698ae4301cb2

Also note the Supabase team is working on a PR https://github.com/supabase/auth/pull/1739 that will allow you to set both email and password at the same time without having to set one first and then the other after. Please don't leave comments on the PR.

**Reply by jkhaui** [reply]
> @jkhaui from your comment you aren't doing what @camwest suggested as you mentioned `signInWithOtp` rather than `updateUser`.
> 
> I've created an example repo here https://github.com/silentworks/supabase_anon_to_user and here is a video attached showing how this works with `updateUser` as @camwest described.
> 
>  Update.Anonymous.User.2024-08-24.083408.mp4 
> Also note the Supabase team is working on a PR [supabase/auth#1739](https://github.com/supabase/auth/pull/1739) that will allow you to set both email and password at the same time without having to set one first and then the other after. Please don't leave comments on the PR.

Ok, thanks a lot for providing this demo. I appreciate it.

I haven't had a chance to try it yet, I will do so soon and let you know if I face any other issues. Cheers

**Reply by zain** [reply]
This worked for me. Thank you @silentworks!
